zapcore/memory_encoder.go

```diff
@@ -45,6 +45,9 @@ func NewMapObjectEncoder() *MapObjectEncoder {
 func (m *MapObjectEncoder) AddArray(key string, v ArrayMarshaler) error {
 	arr := &sliceArrayEncoder{}
 	err := v.MarshalLogArray(arr)
+	if arr.elems == nil {
+		arr.elems = []interface{}{}
```

Comment:
I would make this the default value of `arr.elems` since it'll be appended to anyway.

```
arr := &sliceArrayEncoder{elems: make([]interface{})}
```